### PR TITLE
Add move legality masking in the net.

### PR DESCRIPTION
**Pull request by Rulison**

### CODE REVIEW

File: chess_env.py [hunk 1]
```diff
@@ -72,6 +73,9 @@ def outcome(self, state):
             result = -1
         return result
 
+    def is_legal(self, state, action):
```
Comment (thomas-a-neil) [inline chess_env.py]: Don't think this is used anymore

File: tests/utils.py [hunk 2]
```diff
@@ -52,6 +52,12 @@ def get_legal_actions(self, state):
     def is_game_over(self, state):
         return False
 
+    def get_legal_actions(self, state):
+        return np.array([0, 1])
+
+    def is_game_over(self, state):
```
Comment (thomas-a-neil) [inline tests/utils.py]: looks like these functions are duplicates of lines 49 and 52

File: tests/utils.py [hunk 3]
```diff
@@ -76,9 +82,6 @@ def get_legal_actions(self, state):
     def is_game_over(self, state):
         return False
 
-numline_env = NumlineEnv()
```
Comment (thomas-a-neil) [inline tests/utils.py]: I think this is still used in `test_game.py`? (Jut checked and yeah this is causing some tests to fail)

File: dual_net.py [hunk 4]
```diff
@@ -157,13 +103,93 @@ def __init__(self,
         self.update_op = tf.train.AdamOptimizer(learning_rate).minimize(self.loss)
```
Comment (thomas-a-neil) [inline dual_net.py]: Regarding line 99 above:

```
self.policy_loss = tf.reduce_sum(tf.multiply(self.pi, self.policy_predict))
```

 don't we need a log somewhere?

image

File: dual_net.py [hunk 5]
```diff
@@ -173,10 +199,137 @@ def train(self, boards, pi, z):
...
+
+CHAR_TO_INDEX_MAP = {'K': 0, 'k': 6,
+                     'Q': 1, 'q': 7,
+                     'R': 2, 'r': 8,
+                     'B': 3, 'b': 9,
+                     'N': 4, 'n': 10,
+                     'P': 5, 'p': 11}
+
+
+def state_to_board(state):
```
Comment (thomas-a-neil) [inline dual_net.py]: This is for the full chess environment, right? Not the KQK one? Maybe it deserves its own `ChessEnv` class?

File: dual_net.py [hunk 6]
```diff
@@ -173,10 +199,137 @@ def train(self, boards, pi, z):
...
         self.sess.run([self.update_op], feed_dict={self.board_placeholder: boards,
                                                    self.pi: pi,
                                                    self.z: z})
         loss = self.sess.run([self.loss], feed_dict={self.board_placeholder: boards,
                                                      self.pi: pi,
                                                      self.z: z})
         return loss
+
+# Map from piece to layer in net input.  0-5 are white.
+INDEX_TO_PIECE_MAP = {0: chess.KING, 6: chess.KING,
```
Comment (thomas-a-neil) [inline dual_net.py]: Nice, I'm glad this map is here like this. Makes this code easier to read

File: dual_net.py [hunk 7]
```diff
@@ -173,10 +199,137 @@ def train(self, boards, pi, z):
...
+            for coords in indices:
+                x, y = coords
+                squares.append(8 * (7 - x) + y)
+            for square in squares:
+                pieces[square] = chess.Piece(piece, color)
+        board = chess.Board()
+        board.set_piece_map(pieces)
+        board.turn = int(state[0, 0, 12])
+        return board
+    else:
```
Comment (thomas-a-neil) [inline dual_net.py]: Seems like this has some overlap with the KQK ChessEnv in the other PR. Is our intention to merge them?

File: dual_net.py [hunk 8]
```diff
@@ -157,13 +103,93 @@ def __init__(self,
         self.update_op = tf.train.AdamOptimizer(learning_rate).minimize(self.loss)
         self.sess = sess
 
+
+    def build_model(self,
```
Comment (thomas-a-neil) [inline dual_net.py]: Why did you move this into the class? It doesn't seem to use `self` anywhere

File: dual_net.py [hunk 9]
```diff
@@ -157,13 +103,93 @@ def __init__(self,
...
+                policy_out = conv_block(policy_out, specs)
+            elif specs['layer'] == 'fc':
+                policy_out = layers.flatten(policy_out)
+                policy_out = layers.fully_connected(policy_out,
+                                                    num_outputs=specs['num_outputs'],
+                                                    activation_fn=specs['activation_fn'])
+
+        x = tf.exp(policy_out) * legality_mask_placeholder
+        # Needs reshape to broadcast properly 
+        policy_out = x / tf.reshape(tf.reduce_sum(x, axis=1), shape=((tf.shape(x)[0],) + (1,)))
```
Comment (thomas-a-neil) [inline dual_net.py]: I'm a bit confused as to why we have to do this reshape alchemy..

```
(Pdb) x / tf.reshape(tf.reduce_sum(x, axis=1), shape=((tf.shape(x)[0],) + (1,)))
<tf.Tensor 'truediv:0' shape=(?, 4096) dtype=float32>
(Pdb) x / (tf.reduce_sum(x, axis=1))
<tf.Tensor 'truediv_1:0' shape=(?, 4096) dtype=float32>
```

Looks like that second expression gives us the shape we want. What am I missing?

File: dual_net.py (hunk 4 above)
Comment (Rulison) [reply dual_net.py]: The log is "included" in the net output.  Otherwise we'd be taking logs of negative numbers.

File: chess_env.py (hunk 1 above)
Comment (Rulison) [reply chess_env.py]: There were some merge conflicts in code I wasn't too familiar with.  I erred on the side of redundancy.

File: dual_net.py (hunk 8 above)
Comment (Rulison) [reply dual_net.py]: It used to with another approach I was trying, you're right it doesn't need to be there anymore.

File: dual_net.py (hunk 9 above)
Comment (Rulison) [reply dual_net.py]: Just tried again, when I do the second expression I get:
InvalidArgumentError (see above for traceback): Incompatible shapes: [10,4096] vs. [10]

File: dual_net.py (hunk 5 above)
Comment (Rulison) [reply dual_net.py]: Yeah, I'm leaving it there until we actually make that.

File: dual_net.py (hunk 4 above)
Comment (rajivsambharya) [reply dual_net.py]: Doesn't the self.policy_predict output only a valid probability distribution (sum to 1 and all positive values)? 

x = tf.exp(policy_out) * legality_mask_placeholder

policy_out = x / tf.reshape(tf.reduce_sum(x, axis=1), shape=((tf.shape(x)[0],) + (1,)))

Aside: It's an interesting question whether the model could learn without the log.  I think it might be ok, but may be harder to train.

Comment (Rulison) [reply dual_net.py]: Oh right, we're not using log_softmax anymore, so yes we need to log somewhere

File: dual_net.py [hunk 10]
```diff
@@ -173,10 +196,139 @@ def train(self, boards, pi, z):
         should match input_shape and action_size as set during initialization.
         returns the batch loss
         """
+        self.boards = [state_to_board(board) for board in boards]
```
Comment (thomas-a-neil) [inline dual_net.py]: Okay so this line seems like it makes the net only work on the chess boards (and not extend to TicTacToe, for example). Perhaps store the env as an instance variable and let the env take care of these conversions?

File: dual_net.py (hunk 5 above)
Comment (thomas-a-neil) [reply dual_net.py]: I pushed a rename to make the thing we have now the KQKChessEnv. What's stopping us from moving this out into its own ChessEnv class / module?

File: dual_net.py [hunk 11]
```diff
@@ -162,21 +172,30 @@ def __call__(self, inp):
...
+        for i in range(inp.shape[0]):
+            move_legality_mask[i] = self.env.get_legality_mask(inp[i])
         policy, value = self.sess.run([self.policy_predict, self.value_predict],
-                                      feed_dict={self.board_placeholder: inp})
+                                      feed_dict={self.board_placeholder: inp,
+                                                 self.move_legality_mask: move_legality_mask})
         return policy, value
 
-    def train(self, boards, pi, z):
+    def train(self, states, pi, z, legality_mask):
```
Comment (rajivsambharya) [inline dual_net.py]: If we have the environment as a field of the net then we don't need this legality_mask I think

File: dual_net.py [hunk 12]
```diff
@@ -162,21 +172,30 @@ def __call__(self, inp):
...
+        move_legality_mask = np.zeros(shape=(len(states), self.action_size))
+        for i in range(len(states)):
+            move_legality_mask[i] = self.env.get_legality_mask(states[i])
+        self.sess.run([self.update_op], feed_dict={self.board_placeholder: states,
                                                    self.pi: pi,
-                                                   self.z: z})
-        loss = self.sess.run([self.loss], feed_dict={self.board_placeholder: boards,
+                                                   self.z: z,
+                                                   self.move_legality_mask: move_legality_mask})
+        loss = self.sess.run(self.loss, feed_dict={self.board_placeholder: states,
```
Comment (rajivsambharya) [inline dual_net.py]: We call sess.run twice -- once to update the weights and once to get the loss.  We can do this in 1 swing.

File: tests/test_net.py [hunk 13]
```diff
@@ -1,53 +1,115 @@
...
+        losses = np.zeros(num_iters)
+        for j in range(50):
+            losses[j] = self.net.train(self.states, pi, self.z, legality_mask)
+        self.assertGreater(losses[0], losses[-1])
+
+    def test_train_3_legal_actions(self):
+        '''
+        this is a randomized test
+        on average, the loss should decrease after 50 training steps
+        we assume all actions are legal here
```
Comment (rajivsambharya) [inline tests/test_net.py]: Whooops my bad here.  The assumption is that there are only 3 legal actions

**Review by rajivsambharya — COMMENTED** [review]
Will push a few small changes soon.  I think this looks pretty good!
